### PR TITLE
[LASB-3000] Add mapping for 'unemployed' means.

### PR DESCRIPTION
**Pull request by JHunt16**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/LASB-3000)

Added mapping for 'unemployment' means that we will receive from Crime Apply.

## Checklist

Before you ask people to review this PR:

- [x] Tests should be passing: `./gradlew test`
- [x] Github should not be reporting conflicts; you should have recently run `git rebase main`. **(merged main instead)**
- [x] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [x] You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- [ ] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: crime-applications-adaptor/src/main/resources/schemas/crimeapplicationsadaptor/common/assessmentDetail.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    },
+    "applicantFrequency": {
+      "type": "string",
+      "description": "The Frequency of Applicant",
+      "enum": ["WEEKLY", "2WEEKLY", "4WEEKLY", "MONTHLY", "ANNUALLY"]
+    },
+    "assessmentDetailCode": {
+      "type": "string",
+      "description": "Assessment Detail Code"
+    }
```
Comment (Ross-Nation) [inline crime-applications-adaptor/src/main/resources/schemas/crimeapplicationsadaptor/common/assessmentDetail.json:19]: Should these all be required fields? Will there ever be the case that there is an AssessmentDetails object that is empty?

File: crime-applications-adaptor/src/main/resources/schemas/crimeapplicationsadaptor/common/childWeighting.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    "lowerAgeRange": {
+      "type": "integer",
+      "description": "Lower age range",
+      "enum": [0, 2, 5, 8, 11, 13, 16]
+    },
+    "upperAgeRange": {
+      "type": "integer",
+      "description": "Upper age range",
+      "enum": [1, 4, 7, 10, 12, 15, 18]
+    }
```
Comment (Ross-Nation) [inline crime-applications-adaptor/src/main/resources/schemas/crimeapplicationsadaptor/common/childWeighting.json:20]: Same as above, I think these are all required. If the lower and upper age range aren't filled out then we can't map the noOfChildren

File: crime-applications-adaptor/src/main/resources/schemas/crimeapplicationsadaptor/maat_application_internal.json [hunk 3]
```diff
@@ -153,7 +160,8 @@
     "solicitorAdminEmail",
     "courtCustody",
     "dateCreated",
-    "passported"
+    "passported",
+    "assessment"
```
Comment (Ross-Nation) [inline crime-applications-adaptor/src/main/resources/schemas/crimeapplicationsadaptor/maat_application_internal.json:164]: As I was sort of alluding to in Slack, personally I think we should make this not required. It sounds like Crime Apply won't send an empty means assessment but instead not send the field. For me it's also better if we don't send an assessment with empty means and instead just don't send an assessment object. Happy to discuss this one though. :)

**Review by venkat980 — APPROVED** [review]

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/enums/BenefitDetails.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    public String getValue() {
+        return value;
+    }
+
+    public String getCode() {
+        return code;
+    }
+
+    public static BenefitDetails findByValue(String value) {
+        return Arrays.stream(values()).filter(benefitDetails -> benefitDetails.getValue().equalsIgnoreCase(value)).findFirst().orElseThrow(IllegalArgumentException::new);
```
Comment (lkm0287) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/enums/BenefitDetails.java:35]: Do we want to throw the exception here or at least have it in the method signature?

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java [hunk 5]
```diff
@@ -44,9 +48,29 @@ Applicant map(ClientDetails crimeApplyClientDetails) {
...
         applicant.setPostalAddress(mapAddress(crimeApplyApplicant.getCorrespondenceAddress()));
 
+        mapEmploymentStatus(applicant ,crimeApplyResponse);
+
         return applicant;
     }
 
+    private void mapEmploymentStatus(Applicant applicant, MaatApplicationExternal crimeApplyResponse) {
+        EmploymentStatus employmentStatus = new EmploymentStatus();
+        if (!crimeApplyResponse.getMeansPassport().isEmpty() &&
```
Comment (lkm0287) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java:58]: null check on means passport?

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java [hunk 6]
```diff
@@ -44,9 +48,29 @@ Applicant map(ClientDetails crimeApplyClientDetails) {
...
+                String.valueOf(crimeApplyResponse.getMeansPassport().get(0)).equals(ON_BENEFIT_CHECK)) {
+            employmentStatus.setCode(EmploymentStatus.Code.PASSPORTED);
+        } else if (Objects.nonNull(crimeApplyResponse.getMeansDetails()) && !crimeApplyResponse.getMeansDetails().getIncomeDetails().getEmploymentType().isEmpty()) {
```
Comment (lkm0287) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java:61]: null checks on means details, income details etc?

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapper.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    private static final String UNIVERSAL_CREDIT = "Universal Credit";
+    private static final String JSA = "JSA";
+    public List<AssessmentDetail> mapBenefits(List<Benefit> benefits) {
+        List<AssessmentDetail> assessmentDetails = new ArrayList<>();
+
+        if (Objects.nonNull(benefits)) {
+            for (Benefit benefit : benefits) {
+                AssessmentDetail assessmentDetail = new AssessmentDetail();
+                String benefitType = benefit.getType().value();
+                BenefitDetails benefitDetail = BenefitDetails.findByValue(benefitType);
```
Comment (lkm0287) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapper.java:23]: For example the expected exception is not handled here?

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ChildWeightingMapper.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            map5To7AgeChildren(crimeApplyDependants, childWeightings);
+            map8To10AgeChildren(crimeApplyDependants, childWeightings);
+            map11To12AgeChildren(crimeApplyDependants, childWeightings);
+            map13To15AgeChildren(crimeApplyDependants, childWeightings);
+            map16To18AgeChildren(crimeApplyDependants, childWeightings);
+        }
+        return childWeightings;
+    }
+
+    private void map0To1AgeChildren(List<Dependant> crimeApplyDependants, List<ChildWeighting> childWeightings) {
```
Comment (stone-dj) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ChildWeightingMapper.java:25]: Let's discuss this class and the associated test :)

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/FullMeansAssessmentMapper.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+import javax.validation.constraints.NotNull;
+import java.util.List;
+
+class FullMeansAssessmentMapper {
+
+    private final OutgoingsMapper outgoingsMapper = new OutgoingsMapper();
+
+    @NotNull
+    FullMeansAssessment map(
+            uk.gov.justice.laa.crime.applications.adaptor.model.criminalapplicationsdatastore.general.OutgoingsDetails crimeApplyOutgoingsDetails) {
```
Comment (stone-dj) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/FullMeansAssessmentMapper.java:15]: I don't think we need a fully qualified class name here, we can use an import.

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/FullMeansAssessmentMapper.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+        FullMeansAssessment fullMeansAssessment = new FullMeansAssessment();
+
+        if (crimeApplyOutgoingsDetails == null) {
+            return fullMeansAssessment;
+        }
+        List<AssessmentDetail> assessmentDetails = outgoingsMapper.mapOutgoings(crimeApplyOutgoingsDetails.getOutgoings(), crimeApplyOutgoingsDetails.getHousingPaymentType());
```
Comment (stone-dj) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/FullMeansAssessmentMapper.java:22]: Optional: I would consider extracting out the duplicated method calls into local variables to aid readability.

e.g.
```
        List<Outgoing> outgoings = crimeApplyOutgoingsDetails.getOutgoings();
        Object housingPaymentType = crimeApplyOutgoingsDetails.getHousingPaymentType();

        List<AssessmentDetail> assessmentDetails = outgoingsMapper.mapOutgoings(outgoings, housingPaymentType);
        fullMeansAssessment.setAssessmentDetails(assessmentDetails);
        fullMeansAssessment.setOtherHousingNote(outgoingsMapper.mapOtherHousingFeesNotes(outgoings, housingPaymentType));

```

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OtherIncomeMapper.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+            case FORTNIGHT -> assessmentDetail.setApplicantFrequency( AssessmentDetail.ApplicantFrequency._2_WEEKLY);
+            case FOUR_WEEKS -> assessmentDetail.setApplicantFrequency(AssessmentDetail.ApplicantFrequency._4_WEEKLY);
+            case MONTH -> assessmentDetail.setApplicantFrequency(AssessmentDetail.ApplicantFrequency.MONTHLY);
+            case ANNUAL -> assessmentDetail.setApplicantFrequency(AssessmentDetail.ApplicantFrequency.ANNUALLY);
+        }
+    }
+    public String mapOtherIncomeNotes(List<OtherIncome> otherIncome) {
+        StringBuilder otherNote = new StringBuilder();
+
+        if (Objects.nonNull(otherIncome)) {
```
Comment (stone-dj) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OtherIncomeMapper.java:48]: I would consider having the null check return early to reduce the deep nesting.

```
        if (Objects.isNull(otherIncome)) {
            return StringUtils.EMPTY;
        }
```

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OtherIncomeMapper.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+            for (OtherIncome other : otherIncome) {
+                if (Objects.nonNull(other.getDetails())) {
```
Comment (stone-dj) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OtherIncomeMapper.java:50]: same feedback as above in another mapper with regards to \n and constructing a multiline string

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OutgoingsMapper.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+            case WEEK -> assessmentDetail.setApplicantFrequency(AssessmentDetail.ApplicantFrequency.WEEKLY);
+            case FORTNIGHT -> assessmentDetail.setApplicantFrequency( AssessmentDetail.ApplicantFrequency._2_WEEKLY);
+            case FOUR_WEEKS -> assessmentDetail.setApplicantFrequency(AssessmentDetail.ApplicantFrequency._4_WEEKLY);
+            case MONTH -> assessmentDetail.setApplicantFrequency(AssessmentDetail.ApplicantFrequency.MONTHLY);
+            case ANNUAL -> assessmentDetail.setApplicantFrequency(AssessmentDetail.ApplicantFrequency.ANNUALLY);
+        }
+    }
+
+    public String mapOtherHousingFeesNotes(List<Outgoing> outgoings, Object housingPaymentType) {
+        StringBuilder otherHousingFeesNotes = new StringBuilder();
```
Comment (stone-dj) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OutgoingsMapper.java:54]: same about string construction

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OutgoingsMapper.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+    public List<AssessmentDetail> mapOutgoings(List<Outgoing> outgoings, Object housingPaymentType) {
+        List<AssessmentDetail> assessmentDetails = new ArrayList<>();
+
+        if (Objects.nonNull(outgoings)) {
+            for (Outgoing outgoing : outgoings) {
+                AssessmentDetail assessmentDetail = new AssessmentDetail();
+                String outgoingsType = outgoing.getType().value();
+                // If the type is 'housing', we map against the housing codes instead
+                if (outgoingsType.equals(HOUSING) && Objects.nonNull(housingPaymentType)) {
```
Comment (stone-dj) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OutgoingsMapper.java:26]: I would swap the order of this equality check to aid null safety
`outgoingsType.equals(HOUSING)`
becomes
`HOUSING.equals(outgoingsType)`

File: crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapperTest.java [hunk 15]
```diff
@@ -22,33 +29,78 @@ void setUp() {
...
+        maatApplicationExternal.getMeansDetails().getIncomeDetails().setEmploymentType(employmentType);
+
+        Applicant actualApplicant = applicantMapper.map(maatApplicationExternal);
+
+        assertEquals(EmploymentStatus.Code.SELF, actualApplicant.getEmploymentStatus().getCode());
+    }
+
+    @Test
+    void shouldMapToPassportedIfMeansPassportIsOnBenefitCheck() {
+        MaatApplicationExternal maatApplicationExternal = TestData.getMaatApplication("MaatApplication_unemployed.json");
```
Comment (stone-dj) [inline crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapperTest.java:98]: Good use of test data :)

File: crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapperTest.java [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+import java.math.BigDecimal;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class BenefitsMapperTest {
+
+    private BenefitsMapper benefitsMapper;
+    private final int AMOUNT = 100;
```
Comment (stone-dj) [inline crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapperTest.java:17]: I think these should be static also as the variable names are those of a constant.

File: crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/InitialMeansAssessmentMapperTest.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Test
+    void shouldSuccessfullyMapCrimeApplyIncomeDetailsToInitialMeansAssessment() throws JSONException {
+        IncomeDetails crimeApplyIncomeDetails = TestData.getMaatApplication("MaatApplication_unemployed.json").getMeansDetails().getIncomeDetails();
+
+        InitialMeansAssessment initialMeansAssessment = initialMeansAssessmentMapper.map(crimeApplyIncomeDetails);
+
+        String actualInitialMeansAssessmentJSON = JsonUtils.objectToJson(initialMeansAssessment);
+        String expectedInitialMeansAssessmentJSON = FileUtils.readFileToString("data/expected/crimeapplication/InitialMeansAssessment_mapped.json");
+        JSONAssert.assertEquals(expectedInitialMeansAssessmentJSON, actualInitialMeansAssessmentJSON, JSONCompareMode.STRICT);
```
Comment (stone-dj) [inline crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/InitialMeansAssessmentMapperTest.java:31]: Good use of JSONAssert

File: crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OtherIncomeMapperTest.java [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+import java.math.BigDecimal;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class OtherIncomeMapperTest {
+
+    private OtherIncomeMapper otherIncomeMapper;
+    private final String OTHER_INCOME_DETAILS = "Here are some details.";
+    private final int AMOUNT = 100;
```
Comment (stone-dj) [inline crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OtherIncomeMapperTest.java:18]: Constants should also have the static keyword

File: crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OutgoingsMapperTest.java [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+import java.math.BigDecimal;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class OutgoingsMapperTest {
+    private OutgoingsMapper outgoingsMapper;
+    private final int AMOUNT = 150;
+    private final String DETAILS = "Here are some details about the outgoing.";
```
Comment (stone-dj) [inline crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OutgoingsMapperTest.java:17]: static keyword

**Review by stone-dj — CHANGES_REQUESTED** [review]
Have a look through and let me know of any questions

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/enums/BenefitDetails.java (hunk 4 above)
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/enums/BenefitDetails.java:35]: There is no chance to throw an exception here as we are mapping values from the enum and we have valid data from the Crime Apply data store. there is no chance to get any other values from the CA datastore
As per the functional programming style, no need to add throw exception in the method signature

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java (hunk 5 above)
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java:58]: means passport is a required object, so no need null check.

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapper.java (hunk 7 above)
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapper.java:23]: As mentioned in previous comments, we never throw an exception.

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java (hunk 6 above)
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java:61]: fixed

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java [hunk 20]
```diff
@@ -44,9 +48,29 @@ Applicant map(ClientDetails crimeApplyClientDetails) {
         applicant.setHomeAddress(mapAddress(crimeApplyApplicant.getHomeAddress()));
         applicant.setPostalAddress(mapAddress(crimeApplyApplicant.getCorrespondenceAddress()));
 
+        mapEmploymentStatus(applicant ,crimeApplyResponse);
+
         return applicant;
     }
 
+    private void mapEmploymentStatus(Applicant applicant, MaatApplicationExternal crimeApplyResponse) {
```
[the remark this answers is not in the document]
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java:56]: It is a mapping in Applicant mapper, so we are good with this.

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/enums/BenefitDetails.java [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+
+    public String getValue() {
+        return value;
+    }
+
+    public String getCode() {
+        return code;
+    }
+
+    public static BenefitDetails findByValue(String value) {
```
[the remark this answers is not in the document]
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/enums/BenefitDetails.java:34]: done

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/enums/HousingDetails.java [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+    public String getValue() {
+        return value;
+    }
+
+    public String getCode() {
+        return code;
+    }
+
+    public static HousingDetails findByValue(String value) {
+        return Arrays.stream(values()).filter(housingDetails -> housingDetails.getValue().equalsIgnoreCase(value)).findFirst().orElseThrow(IllegalArgumentException::new);
```
[the remark this answers is not in the document]
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/enums/HousingDetails.java:27]: done

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/enums/OtherIncomeDetails.java [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+    public String getValue() {
+        return value;
+    }
+
+    public String getCode() {
+        return code;
+    }
+
+    public static OtherIncomeDetails findByValue(String value) {
+        return Arrays.stream(values()).filter(otherIncomeDetails -> otherIncomeDetails.getValue().equalsIgnoreCase(value)).findFirst().orElseThrow(IllegalArgumentException::new);
```
[the remark this answers is not in the document]
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/enums/OtherIncomeDetails.java:37]: done

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/enums/BenefitDetails.java (hunk 4 above)
Comment (lkm0287) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/enums/BenefitDetails.java:35]: I'm talking about this part: .orElseThrow(IllegalArgumentException::new)

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java [hunk 24]
```diff
@@ -2,27 +2,31 @@
...
+import uk.gov.justice.laa.crime.applications.adaptor.model.criminalapplicationsdatastore.general.EmploymentType;
 
 import javax.validation.constraints.NotNull;
+import java.util.Objects;
 
 class ApplicantMapper {
 
     private static final boolean APPLICANT_HAS_PARTNER_DEFAULT_FALSE = false;
     private static final String PARTNER_CONTRARY_INTEREST_CODE_DEFAULT_NOCON = "NOCON";
+    public static final String ON_BENEFIT_CHECK = "on_benefit_check";
```
[the remark this answers is not in the document]
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java:17]: done

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ApplicantMapper.java:56]: done

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapper.java [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+                assessmentDetail.setAssessmentDetailCode(benefitDetail.getCode());
+                assessmentDetail.setApplicantAmount(new BigDecimal(benefit.getAmount()));
+                mapFrequency(assessmentDetail, benefit.getFrequency());
+                assessmentDetails.add(assessmentDetail);
+            }
+        }
+        return assessmentDetails;
+    }
+
+    private void mapFrequency(AssessmentDetail assessmentDetail, Benefit.Frequency frequency) {
```
[the remark this answers is not in the document]
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapper.java:33]: done

File: crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OutgoingsMapperTest.java (hunk 19 above)
Comment (nittagr) [reply crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OutgoingsMapperTest.java]: done

File: crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OtherIncomeMapperTest.java (hunk 18 above)
Comment (nittagr) [reply crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OtherIncomeMapperTest.java]: done

File: crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapperTest.java (hunk 16 above)
Comment (nittagr) [reply crime-applications-adaptor/src/test/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapperTest.java]: done

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapper.java [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+            case ANNUAL -> assessmentDetail.setApplicantFrequency(AssessmentDetail.ApplicantFrequency.ANNUALLY);
+        }
+    }
+
+    public String mapOtherBenefitNotes(List<Benefit> benefits) {
+        StringBuilder otherNote = new StringBuilder();
+        if (Objects.nonNull(benefits)) {
+            for (Benefit benefit : benefits) {
+                if (Objects.nonNull(benefit.getDetails())) {
+                    otherNote.append("\n");
```
[the remark this answers is not in the document]
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapper.java]: done

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapper.java [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+            case WEEK -> assessmentDetail.setApplicantFrequency(AssessmentDetail.ApplicantFrequency.WEEKLY);
+            case FORTNIGHT -> assessmentDetail.setApplicantFrequency( AssessmentDetail.ApplicantFrequency._2_WEEKLY);
+            case FOUR_WEEKS -> assessmentDetail.setApplicantFrequency(AssessmentDetail.ApplicantFrequency._4_WEEKLY);
+            case MONTH -> assessmentDetail.setApplicantFrequency(AssessmentDetail.ApplicantFrequency.MONTHLY);
+            case ANNUAL -> assessmentDetail.setApplicantFrequency(AssessmentDetail.ApplicantFrequency.ANNUALLY);
+        }
+    }
+
+    public String mapOtherBenefitNotes(List<Benefit> benefits) {
+        StringBuilder otherNote = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/BenefitsMapper.java]: done

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/FullMeansAssessmentMapper.java (hunk 9 above)
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/FullMeansAssessmentMapper.java]: done

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OtherIncomeMapper.java (hunk 11 above)
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OtherIncomeMapper.java]: done

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OutgoingsMapper.java (hunk 13 above)
Comment (nittagr) [reply crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/OutgoingsMapper.java]: done

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ChildWeightingMapper.java [hunk 28]
```diff
@@ -11,62 +11,21 @@ public class ChildWeightingMapper {
...
-            childWeightings.add(new ChildWeighting(age13to15, ChildWeighting.LowerAgeRange._13, ChildWeighting.UpperAgeRange._15));
         }
+        return childWeightings;
     }
-    private void map16To18AgeChildren(List<Dependant> crimeApplyDependants, List<ChildWeighting> childWeightings) {
-        int age16to18 = Math.toIntExact(crimeApplyDependants.stream().filter(dependant -> (dependant.getAge() >= 16 && dependant.getAge() <= 18)).count());
-        if (age16to18 > 0) {
-            childWeightings.add(new ChildWeighting(age16to18, ChildWeighting.LowerAgeRange._16, ChildWeighting.UpperAgeRange._18));
+    private void mapChildrenBasedOnAgeRange(List<Dependant> crimeApplyDependants, List<ChildWeighting> childWeightings, ChildWeighting.LowerAgeRange lowerAgeRange, ChildWeighting.UpperAgeRange upperAgeRange) {
+        int numberOfChildren = Math.toIntExact(crimeApplyDependants.stream().filter(dependant -> (dependant.getAge() >= lowerAgeRange.value() && dependant.getAge() <= upperAgeRange.value())).count());
```
Comment (stone-dj) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/ChildWeightingMapper.java]: I didn't notice this before, how many children are we expecting !? lol
`Math.toIntExact`
Agreed, better to safe than sorry! :)

File: crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/FrequencyMapper.java [hunk 29]
```diff
@@ -0,0 +1,38 @@
+package uk.gov.justice.laa.crime.applications.adaptor.mapper.crimeapply;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import uk.gov.justice.laa.crime.applications.adaptor.model.crimeapplicationsadaptor.common.AssessmentDetail.ApplicantFrequency;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
```
Comment (stone-dj) [inline crime-applications-adaptor/src/main/java/uk/gov/justice/laa/crime/applications/adaptor/mapper/crimeapply/FrequencyMapper.java:7]: can use this annotation @UtilityClass

**Review by stone-dj — APPROVED** [review]
Changes look good

**Review by nittagr — APPROVED** [review]

**Pull request merged**